batch/batch/driver/instance_collection/pool.py

```diff
@@ -603,51 +611,62 @@ async def schedule_loop_body(self):
+            async for job_group in self.db.select_and_fetchall(
                 """
-SELECT batches.id, job_groups_cancelled.id IS NOT NULL AS cancelled, userdata, user, format_version
-FROM batches
-LEFT JOIN job_groups_cancelled
-       ON batches.id = job_groups_cancelled.id
-WHERE user = %s AND `state` = 'running';
+SELECT job_groups.batch_id, job_groups.job_group_id, t.cancelled IS NOT NULL AS cancelled, userdata, job_groups.user, format_version
+FROM job_groups
+LEFT JOIN batches ON job_groups.batch_id = batches.id
+LEFT JOIN LATERAL (
+  SELECT 1 AS cancelled
+  FROM job_group_self_and_ancestors
+  INNER JOIN job_groups_cancelled
+    ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
+      job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
+  WHERE job_groups.batch_id = job_group_self_and_ancestors.batch_id AND
+    job_groups.job_group_id = job_group_self_and_ancestors.job_group_id
+) AS t ON TRUE
+WHERE job_groups.user = %s AND job_groups.`state` = 'running'
+ORDER BY job_groups.batch_id, job_groups.job_group_id;
 """,
                 (user,),
                 "user_runnable_jobs__select_running_batches",
             ):
                 async for record in self.db.select_and_fetchall(
                     """
-SELECT jobs.job_id, spec, cores_mcpu, regions_bits_rep, time_ready
+SELECT jobs.job_id, spec, cores_mcpu, regions_bits_rep, time_ready, job_group_id
 FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
 LEFT JOIN jobs_telemetry ON jobs.batch_id = jobs_telemetry.batch_id AND jobs.job_id = jobs_telemetry.job_id
-WHERE jobs.batch_id = %s AND inst_coll = %s AND jobs.state = 'Ready' AND always_run = 1
-ORDER BY jobs.batch_id, inst_coll, state, always_run, -n_regions DESC, regions_bits_rep, jobs.job_id
+WHERE jobs.batch_id = %s AND job_group_id = %s AND inst_coll = %s AND jobs.state = 'Ready' AND always_run = 1
+ORDER BY jobs.batch_id, jobs.job_group_id, inst_coll, state, always_run, -n_regions DESC, regions_bits_rep, jobs.job_id
 LIMIT 300;
 """,
-                    (batch['id'], self.pool.name),
+                    (job_group['batch_id'], job_group['job_group_id'], self.pool.name),
                     "user_runnable_jobs__select_ready_always_run_jobs",
                 ):
-                    record['batch_id'] = batch['id']
-                    record['userdata'] = batch['userdata']
-                    record['user'] = batch['user']
-                    record['format_version'] = batch['format_version']
+                    record['batch_id'] = job_group['batch_id']
+                    record['job_group_id'] = job_group['job_group_id']
+                    record['userdata'] = job_group['userdata']
+                    record['user'] = job_group['user']
+                    record['format_version'] = job_group['format_version']
                     yield record
-                if not batch['cancelled']:
+                if not job_group['cancelled']:
```

Comment:
Continuing form https://github.com/hail-is/hail/pull/14170/files#r1476849027

We can leave as is for now, but let's make an issue. This code is more difficult to read than it need be. The simplification in this case is exactly the one from job_private:
```python3
...
AND (always_run OR NOT %s)
...
"""
, (..., job_group['cancelled'])
```